### DISCUSSION TITLE
Couldn't we just store Instagram image CDN links instead of caching images?

### DISCUSSION THREAD

**Discussion by codelif** [codeblech/jpgram#12] [Ideas]
## What?
If we use instaloader without the args `--no-metadata-json`, instaloader will also store a json file with some metadata like `display_url` which contains the direct links to Instagram's CDN.

For example, 
![image](https://github.com/user-attachments/assets/3538517b-d2f4-495a-b5ef-8e324b054d3f)

This link points to the post's image and can be accessed and streamed directly:
![image](https://github.com/user-attachments/assets/8529fe4d-85d6-4c16-87df-d13b91c08344)

## Why bother?
 - Using this along with `--no-pictures` argument would save a lot of storage in the long term. (Repo size has already crossed 550 MBs)
 - Instagram's CDN would be faster.
 - Instaloader would also take less time because it doesn't have to cache the images.
 - Metadata JSON also contains links for smaller-sized images and thumbnails which can be used to load the page faster in the future.

## How much extra work would it be?
Not much. In terms of changes to the CD pipeline, just changing the args of instaloader would be enough.

For the code, for every post it would have to parse a JSON file to fetch the image URL.

I can create a proto today/tomorrow (if you want).

## Potential Caveats
 - Would make the project dependent on IG's CDN. 
 - If they switch CDNs, would need to update metadata cache as well.

**Comment by codeblech** [comment]
i see u fw this project
great effort brotha😎 
the cdn thing...i was planning to do that (but fogor thanks for reminding)

> If they switch CDNs, would need to update     metadata cache as well.

do they do that regularly/frequently?

if you wanna go right ahead with a PR that'd be awesome 🔥
else I'd be on it anyways 🦾

**Reply by codeblech** [reply]
I'm opening an issue to track this

**Comment by codelif** [comment]
I am not sleeping tonight. Expect a PR when you wake up.

**Reply by codeblech** [reply]
😈
![Screenshot_20240819-000445.png](https://github.com/user-attachments/assets/cb63a5bf-8e7b-4010-894b-a1b0f4cbb350)

**Reply by codeblech** [reply]
gh notifications not working for me
hit me up on telegram/whatsapp if you need to ask anything urgent (though mention it on gh as well for the sake of open source)

**Comment by codelif** [comment]
> i see u fw this project

I do.


> do they do that regularly/frequently?

Not sure, probably not daily. I expect atleast 1 month (though I am pulling this out of my a**)

**Comment by codelif** [comment]
So..... The proto is done but.... it won't work.
The IG CDN has disabled Crossorigin requests. So it only allows requests from instagram.com 

Sadly, I don't think we can fix this. All browsers support the CORS policies so this will be enforced on any W3C-compatible browser (which includes all of them)


https://github.com/user-attachments/assets/c0d244a5-2b56-4f86-92db-6dab8f71d1ba

We could potentially make a "proxy server" which would stream these files for us, but that would induce too much latency.

Please suggest some solution.

**Reply by codeblech** [reply]
i see i see
if i remember correctly the magazine/static/magazine/ folder is ~172MB
we probably need to remove the assets from version control, that'd do a lot of saving 

as for the cdn thing, try looking at how apps/wrappers like instapro & instander do it

also, when you open a single cdn link from your browser, why does that work?

**Reply by codelif** [reply]
> also, when you open a single cdn link from your browser, why does that work?

Because the browser is directly accessing that particular URI, in the other case the webpage (127.0.0.1:8000) asks the browser to load the image, but the browser rejects because the target server (fbcdn.net) has denied CORS to the source origin (127.0.0.1:8000).

This is an artificial limitation set by Browsers (known as CORS), so any website couldn't just send any malicious requests to any domain (especially using your own browser, imagine cookies). If this did not exist, we could potentially send any network requests using the client's browser (imagine any website could access our mail accounts logged in our browser) 

This also exists so websites (like ours) can't just scrape data right off any other website using your browser. 

CORS gives control to a website for which other websites can access it.

All browsers (at least the relevant ones) have this artificial limitation. 

For more info: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

> as for the cdn thing, try looking at how apps/wrappers like instapro & instander do it

They are directly sending requests through python rather than asking the browser to do it so there is no CORS to stop them.

**Reply by codeblech** [reply]
aight aight
can you think of a solution?
anyways, don't stress too much on it
we can find other ways of reducing storage usage, maybe?

**Reply by codeblech** [reply]
also, a bigger problem is running instaloader on the prod server. i tried it once, and all requests failed. I haven't yet figured out why.
if you like, you can work on this issue #10

on top of that, we need to setup a task on python anywhere that runs the instaloader command daily
i dont know if they allow cron jobs. that'd do the job too

**Reply by codelif** [reply]
> can you think of a solution?

no, not currently. Some way or other, it leads to the server downloading the image.

> we can find other ways of reducing storage usage, maybe?

Hopefully...... 
Compression maybe.

**Reply by codeblech** [reply]
I already use compression 
I'm using cwebp to compress to webp

**Reply by codeblech** [reply]
i forgot to mention; that too, needs to be setup as an automated task
